src/api.h

```diff
@@ -39,25 +34,325 @@ AWS_CRT_API const char *aws_crt_error_debug_str(int);
 AWS_CRT_API void aws_crt_reset_error(void);
 
 /* IO */
-AWS_CRT_API struct aws_event_loop_group *
-aws_crt_event_loop_group_new(uint16_t max_threads);
-AWS_CRT_API void
-aws_crt_event_loop_group_release(struct aws_event_loop_group *elg);
+typedef struct aws_event_loop_group aws_crt_event_loop_group;
+typedef struct _aws_crt_event_loop_group_options aws_crt_event_loop_group_options;
+AWS_CRT_API aws_crt_event_loop_group_options *aws_crt_event_loop_group_options_new(void);
+AWS_CRT_API void aws_crt_event_loop_group_options_release(aws_crt_event_loop_group_options *options);
+AWS_CRT_API void aws_crt_event_loop_group_options_set_max_threads(
+    aws_crt_event_loop_group_options *options,
+    uint16_t max_threads);
+AWS_CRT_API aws_crt_event_loop_group *aws_crt_event_loop_group_new(aws_crt_event_loop_group_options *options);
+AWS_CRT_API void aws_crt_event_loop_group_release(aws_crt_event_loop_group *elg);
+
+/* Input stream */
+typedef struct aws_input_stream aws_crt_input_stream;
+typedef struct _aws_crt_input_stream_options aws_crt_input_stream_options;
+typedef struct aws_stream_status aws_crt_stream_status;
+typedef int(aws_crt_input_stream_seek_fn)(void *user_data, int64_t offset, int seek_basis);
+typedef int(aws_crt_input_stream_read_fn)(void *user_data, uint8_t *dest, size_t dest_length);
+typedef int(aws_crt_input_stream_get_status_fn)(void *user_data, aws_crt_stream_status *status);
+typedef int(aws_crt_input_stream_get_length_fn)(void *user_data, int64_t *out_length);
+typedef int(aws_crt_input_stream_destroy_fn)(void *user_data);
+AWS_CRT_API aws_crt_input_stream_options *aws_crt_input_stream_options_new(void);
```

Comment:
input_stream is going to be a case where the C layer needs to keep an external object alive.

so there needs to be a callback that lets the external object know when it's safe to die.
not sure if the callback and userdata should be part of the new() call, or separate setters